File: java/src/main/java/com/aerospike/helper/query/Qualifier.java
package com.aerospike.helper.query;

import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Set;

import com.aerospike.client.Value;
import com.aerospike.client.command.ParticleType;
import com.aerospike.client.query.Filter;
/**
 * Generic Bin qualifier
 * @author peter
 *
 */
public class Qualifier implements Map<String, Object>{
	private static final String FIELD = "field";
	private static final String VALUE2 = "value2";
	private static final String VALUE1 = "value1";
	private static final String OPERATION = "operation";
	private Map<String, Object> internalMap;
	public enum FilterOperation {
		EQ, GT, GTEQ, LT, LTEQ, NOTEQ, BETWEEN, START_WITH, ENDS_WITH
	}

	public Qualifier() {
		super();
		internalMap = new HashMap<String, Object>();
	}
	public Qualifier(String field, FilterOperation operation, Value value1) {
		this();
		internalMap.put(FIELD, field);
		internalMap.put(OPERATION, operation);
		internalMap.put(VALUE1, value1);
	}
	public Qualifier(String field, FilterOperation operation, Value value1, Value value2) {
		this(field, operation, value1);
		internalMap.put(VALUE2, value2);
	}

	public FilterOperation getOperation(){
		return (FilterOperation) internalMap.get(OPERATION);
	}
	public String getField(){
		return (String) internalMap.get(FIELD);
	}
	public Value getValue1(){
		return (Value) internalMap.get(VALUE1);
	}
	public Value getValue2(){
		return (Value) internalMap.get(VALUE2);
	}

	@SuppressWarnings("deprecation")
	public Filter asFilter(){
		FilterOperation op = getOperation();
		switch (op) {
		case EQ:
			return Filter.equal(getField(), getValue1());
		case BETWEEN:
			return Filter.range(getField(), getValue1(), getValue2());
		default:
			return null;
		}
	}

	public String luaFilterString(){
		String value1 = luaValueString(getValue1());
		FilterOperation op = getOperation();
		switch (op) {
		case EQ:
			return String.format("%s == %s", luaFieldString(getField()),  value1);
		case NOTEQ:
			return String.format("%s ~= %s", luaFieldString(getField()), value1);
		case GT:
			return String.format("%s > %s", luaFieldString(getField()), value1);
		case GTEQ:
			return String.format("%s >= %s", luaFieldString(getField()), value1);
		case LT:
			return String.format("%s < %s", luaFieldString(getField()), value1);
		case LTEQ:
			return String.format("%s <= %s", luaFieldString(getField()), value1);
		case BETWEEN:
			String value2 = luaValueString(getValue2());
			String fieldString = luaFieldString(getField()); 
			return String.format("%s >= %s and %s <= %s  ", fieldString, value1, luaFieldString(getField()), value2);
		case START_WITH:
			return String.format("string.sub(%s,1,string.len(%s))==%s", luaFieldString(getField()), value1, value1);			
		case ENDS_WITH:
			return String.format("%s=='' or string.sub(%s,-string.len(%s))==%s", 
					value1,
					luaFieldString(getField()),
					value1,
					value1);			
		}
		return "";
	}
	
	protected String luaFieldString(String field){
		return String.format("rec['%s']", field);
	}

	protected String luaValueString(Value value){
		String res = null;
		int type = value.getType();
		switch (type) {
//		case ParticleType.LIST:
//			res = value.toString();
//			break;
//		case ParticleType.MAP:
//			res = value.toString();
//			break;
//		case ParticleType.DOUBLE:
//			res = value.toString();
//			break;
		case ParticleType.STRING:
			res = String.format("'%s'", value.toString());
			break;
		default:
			res = value.toString();
			break;
		}
		return res;
	}



	/*
	 * (non-Javadoc)
	 * @see java.util.Map#size()
	 */
	@Override
	public int size() {
		return internalMap.size();
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#isEmpty()
	 */
	@Override
	public boolean isEmpty() {
		return internalMap.isEmpty();
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#containsKey(java.lang.Object)
	 */
	@Override
	public boolean containsKey(java.lang.Object key) {
		return internalMap.containsKey(key);
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#containsValue(java.lang.Object)
	 */
	@Override
	public boolean containsValue(java.lang.Object value) {
		return internalMap.containsValue(value);
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#get(java.lang.Object)
	 */
	@Override
	public Object get(java.lang.Object key) {
		return internalMap.get(key);
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#put(java.lang.Object, java.lang.Object)
	 */
	@Override
	public Object put(String key, Object value) {
		return internalMap.put(key, value);
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#remove(java.lang.Object)
	 */
	@Override
	public Object remove(java.lang.Object key) {
		return internalMap.remove(key);
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#putAll(java.util.Map)
	 */
	@Override
	public void putAll(Map<? extends String, ? extends Object> m) {
		internalMap.putAll(m);
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#clear()
	 */
	@Override
	public void clear() {
		internalMap.clear();
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#keySet()
	 */
	@Override
	public Set<String> keySet() {
		return internalMap.keySet();
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#values()
	 */
	@Override
	public Collection<Object> values() {
		return internalMap.values();
	}
	/*
	 * (non-Javadoc)
	 * @see java.util.Map#entrySet()
	 */
	@Override
	public Set<java.util.Map.Entry<String, Object>> entrySet() {
		return internalMap.entrySet();
	}

	@Override
	public String toString() {
		String output = String.format("%s:%s:%s:%s",getField(), getOperation(), getValue1(), getValue2());
		return output;
	}
}

File: csharp/AerospikeHelper/AerospikeHelper/Query/Qualifier.cs
using System;
using System.Collections.Generic;
using System.Collections;
using Aerospike.Client;

namespace AerospikeHelper.Query
{
	public class Qualifier : IDictionary<String, Object> 
	{
		private const String FIELD = "field";
		private const String VALUE2 = "value2";
		private const String VALUE1 = "value1";
		private const String OPERATION = "operation";
		private const String COLLECTION_OPERATION = "collection-operation";
		protected Dictionary<String, Object> internalMap;
		public enum FilterOperation {
			EQ, GT, GTEQ, LT, LTEQ, NOTEQ, BETWEEN, START_WITH, ENDS_WITH, 
			LIST_CONTAINS, MAP_KEYS_CONTAINS, MAP_VALUES_CONTAINS, 
			LIST_BETWEEN, MAP_KEYS_BETWEEN, MAP_VALUES_BETWEEN
		}

		public Qualifier() {
			internalMap = new Dictionary<String, Object>();
		}
		public Qualifier(String field, FilterOperation operation, Value value1) : this() {

			internalMap[FIELD] = field;
			internalMap[OPERATION] = operation;
			internalMap[VALUE1] = value1;
		}
		public Qualifier(String field, FilterOperation operation, Value value1, Value value2) : this(field, operation, value1) {
			internalMap[VALUE2] = value2;
		}



		public FilterOperation getOperation(){
			return (FilterOperation) internalMap[OPERATION];
		}
		public String Field {
			get {
				return (String)internalMap [FIELD];
			}
		}
		public Value Value1{
			get {
				return (Value)internalMap [VALUE1];
			}
		}
		public Value Value2{
			get {
				return (Value)internalMap [VALUE2];
			}
		}

		public Filter AsFilter(){
			FilterOperation op = getOperation();
			switch (op) {
			case FilterOperation.EQ:
				return BinEquality ();
			case FilterOperation.BETWEEN:
				return BinRange ();
			case FilterOperation.LIST_CONTAINS:
				return collectionContains(IndexCollectionType.LIST);
			case FilterOperation.MAP_KEYS_CONTAINS:
				return collectionContains(IndexCollectionType.MAPKEYS);
			case FilterOperation.MAP_VALUES_CONTAINS:
				return collectionContains(IndexCollectionType.MAPVALUES);
			case FilterOperation.LIST_BETWEEN:
				return collectionRange(IndexCollectionType.LIST);
			case FilterOperation.MAP_KEYS_BETWEEN:
				return collectionRange(IndexCollectionType.MAPKEYS);
			case FilterOperation.MAP_VALUES_BETWEEN:
				return collectionRange(IndexCollectionType.MAPKEYS);
			default:
				return null;
			}
		}

		private Filter BinRange(){
			return Filter.Range(Field, Value1.ToLong(), Value2.ToLong());
		}

		private Filter BinEquality(){
			int valType = Value1.Type;
			switch (valType){
			case ParticleType.INTEGER:
				return Filter.Equal(Field, Value1.ToLong());
			case ParticleType.STRING:
				return Filter.Equal(Field, Value1.ToString());
			}
			return null;
		}

		private Filter collectionContains(IndexCollectionType collectionType){
			Value val = Value1;
			int valType = val.Type;
			switch (valType){
			case ParticleType.INTEGER:
				return Filter.Contains(Field, collectionType, val.ToLong());
			case ParticleType.STRING:
				return Filter.Contains(Field, collectionType, val.ToString());
			}
			return null;
		}
		private Filter collectionRange(IndexCollectionType collectionType){
			return Filter.Range(Field, collectionType, Value1.ToLong(), Value2.ToLong());
		}

		public String luaFilterString(){
			String value1 = luaValueString(Value1);
			FilterOperation op = getOperation();
			switch (op) {
			case FilterOperation.EQ:
				return String.Format("%s == %s", luaFieldString(Field),  value1);
			case FilterOperation.LIST_CONTAINS:
				return String.Format("containsValue(%s, %s)", luaFieldString(Field),  value1);
			case FilterOperation.MAP_KEYS_CONTAINS:
				return String.Format("containsKey(%s, %s)", luaFieldString(Field),  value1);
			case FilterOperation.MAP_VALUES_CONTAINS:
				return String.Format("containsValue(%s, %s)", luaFieldString(Field),  value1);
			case FilterOperation.NOTEQ:
				return String.Format("%s ~= %s", luaFieldString(Field), value1);
			case FilterOperation.GT:
				return String.Format("%s > %s", luaFieldString(Field), value1);
			case FilterOperation.GTEQ:
				return String.Format("%s >= %s", luaFieldString(Field), value1);
			case FilterOperation.LT:
				return String.Format("%s < %s", luaFieldString(Field), value1);
			case FilterOperation.LTEQ:
				return String.Format("%s <= %s", luaFieldString(Field), value1);
			case FilterOperation.BETWEEN:
				String value2 = luaValueString(Value2);
				String fieldString = luaFieldString(Field); 
				return String.Format("%s >= %s and %s <= %s  ", fieldString, value1, luaFieldString(Field), value2);
			case FilterOperation.LIST_BETWEEN:
				value2 = luaValueString(Value2);
				return String.Format("rangeValue(%s, %s, %s)", luaFieldString(Field),  value1, value2);
			case FilterOperation.MAP_KEYS_BETWEEN:
				value2 = luaValueString(Value2);
				return String.Format("rangeKey(%s, %s, %s)", luaFieldString(Field),  value1, value2);
			case FilterOperation.MAP_VALUES_BETWEEN:
				value2 = luaValueString(Value2);
				return String.Format("rangeValue(%s, %s, %s)", luaFieldString(Field),  value1, value2);
			case FilterOperation.START_WITH:
				return String.Format("string.sub(%s,1,string.len(%s))==%s", luaFieldString(Field), value1, value1);			
			case FilterOperation.ENDS_WITH:
				return String.Format("%s=='' or string.sub(%s,-string.len(%s))==%s", 
					value1,
					luaFieldString(Field),
					value1,
					value1);			
			}
			return "";
		}

		protected String luaFieldString(String field){
			return String.Format("rec['%s']", field);
		}

		protected String luaValueString(Value value){
			String res = null;
			int type = value.Type;
			switch (type) {
			//		case ParticleType.LIST:
			//			res = value.toString();
			//			break;
			//		case ParticleType.MAP:
			//			res = value.toString();
			//			break;
			//		case ParticleType.DOUBLE:
			//			res = value.toString();
			//			break;
			case ParticleType.STRING:
				res = String.Format("'%s'", value.ToString());
				break;
			default:
				res = value.ToString();
				break;
			}
			return res;
		}
		#region IDictionary Members
		public bool IsReadOnly { get { return false; } }
		public bool ContainsKey(string key)
		{
			return internalMap.ContainsKey(key);
		}
		public bool IsFixedSize { get { return false; } }

		public bool Remove(string key)
		{
			return internalMap.Remove (key);
		}
		public bool Remove(object key)
		{
			return internalMap.Remove ((String)key);
		}
		public void Clear() { internalMap.Clear(); }

		public void Add(string key, object value) 
		{
			internalMap.Add (key, value);
		}

		public void Add(object key, object value) 
		{
			internalMap.Add ((String)key, value);
		}

		public ICollection<string> Keys
		{
			get
			{
				return internalMap.Keys;
			}
		}
		public ICollection<object> Values
		{
			get
			{
				return internalMap.Values;
			}
		}
		public object this[object key]
		{
			get
			{   
				return internalMap [(String)key];
			}

			set
			{
				internalMap [(String)key] = value;
			}
		}
		public object this[string key]
		{
			get
			{   
				return internalMap [key];
			}

			set
			{
				internalMap [key] = value;
			}
		}
		private Boolean TryGetIndexOfKey(Object key, out Object index)
		{
			return internalMap.TryGetValue((string)key, out index);
		}
		public IEnumerator GetEnumerator()
		{
			return internalMap.GetEnumerator();
		}

		public bool  TryGetValue(string key, out object value){
			return internalMap.TryGetValue (key, out value);
		}

		public int Count {
			get {
				return internalMap.Count;
			}
		}

		#endregion

	}
}

